syz-manager/covfilter.go

```diff
@@ -0,0 +1,215 @@
+		if err != nil {
+			return fmt.Errorf("failed to open raw PCs file: %v", err)
+		}
+		for {
+			var encode uint64
+			_, err := fmt.Fscanf(rawFile, "0x%x\n", &encode)
+			if err != nil {
+				break
+			}
+			pc := uint32(encode & 0xffffffff)
+			weight := float32((encode >> 32) & 0xffff)
+			covFilter.weightedPCs[pc] = weight
+		}
+		rawFile.Close()
+	}
+
+	enabledFuncs := make(map[string]bool)
+	enabledFiles := make(map[string]bool)
+	if len(filesRegexp) > 0 || len(funcsRegexp) > 0 {
+		if reportGenerator == nil {
+			return fmt.Errorf("reportGenerator used without initialization")
+		}
+		pcs := make([]uint64, 0)
+		symNames, cuNames, symPCs := reportGenerator.GetSymbolsInfo()
+		for i, symName := range symNames {
+			for _, r := range funcsRegexp {
+				if ok := r.MatchString(symName); ok {
+					enabledFuncs[symName] = true
+					pcs = append(pcs, symPCs[i]...)
+				}
+			}
+			for _, r := range filesRegexp {
+				if ok := r.MatchString(cuNames[i]); ok {
+					enabledFuncs[symName] = true
+					enabledFiles[cuNames[i]] = true
+					pcs = append(pcs, symPCs[i]...)
+				}
+			}
+		}
+		for _, pc := range pcs {
+			covFilter.weightedPCs[uint32(pc)] = 1.0
+		}
+	}
+	for funcs := range enabledFuncs {
+		log.Logf(1, "enabled kernel function: %s", funcs)
+	}
+	for files := range enabledFiles {
+		log.Logf(1, "enabled kernel file: %s", files)
+	}
+	return nil
+}
+
+func getRegexps(regexpStrings []string) ([]regexp.Regexp, error) {
+	var regexps []regexp.Regexp
+	for _, rs := range regexpStrings {
+		r, err := regexp.Compile(rs)
+		if err != nil {
+			return nil, fmt.Errorf("failed to compile regexp: %v", err)
+		}
+		regexps = append(regexps, *r)
+	}
+	return regexps, nil
+}
+
+func (covFilter *CoverFilter) createBitmap() error {
+	covFilter.detectRegion()
+	if covFilter.pcSize > 0 {
+		log.Logf(0, "coverage filter from 0x%x to 0x%x, size 0x%x",
+			covFilter.pcStart, covFilter.pcEnd, covFilter.pcSize)
+	} else {
+		return fmt.Errorf("coverage filter is enabled but nothing will be filtered")
+	}
+
+	bitmapFile, err := os.OpenFile(covFilter.bitmapFilename, os.O_RDWR|os.O_CREATE|os.O_TRUNC, 0644)
+	if err != nil {
+		return fmt.Errorf("failed to open or create bitmap: %s", err)
+	}
+	defer bitmapFile.Close()
+	bitmap := covFilter.bitmapBytes()
+	_, err = bitmapFile.Write(bitmap)
+	if err != nil {
+		return fmt.Errorf("failed to write bitmap: %s", err)
+	}
+	return nil
+}
+
+func (covFilter *CoverFilter) detectRegion() {
+	covFilter.pcStart = 0xffffffff
+	covFilter.pcEnd = 0x0
+	for pc := range covFilter.weightedPCs {
+		if pc < covFilter.pcStart {
+			covFilter.pcStart = pc
+		}
+		if pc > covFilter.pcEnd {
+			covFilter.pcEnd = pc
+		}
+	}
+	// align
+	covFilter.pcStart &= 0xfffffff0
+	covFilter.pcEnd |= 0xf
+	covFilter.pcEnd++
+	covFilter.pcSize = 0
+	if covFilter.pcStart < covFilter.pcEnd {
+		covFilter.pcSize = covFilter.pcEnd - covFilter.pcStart
+	} else {
+		covFilter.pcSize = 0
+	}
+}
+
+func (covFilter *CoverFilter) bitmapBytes() []byte {
```

Comment:
These 2 functions can be made a bit shorter:
```
func (covFilter *CoverFilter) bitmapBytes() []byte {
	// The file starts with two uint32: covFilterStart and covFilterSize,
	// and a bitmap with size ((covFilterSize>>4) + 7)/8 bytes follow them.
	// 8-bit = 1-byte, additional 1-byte to prevent overflow
	data := make([]byte, ((covFilter.pcSize >> 4) + 7) / 8 + 8)
	order := binary.BigEndian
	if covFilter.target.LittleEndian {
		order = binary.LittleEndian
	}
	order.PutUint32(data, covFilter.pcStart)
	order.PutUint32(data[4:], covFilter.pcSize)
	bitmap = data[8:]
	for pc := range covFilter.weightedPCs {
		// The lowest 4-bit is dropped,
		pc = (pc - covFilter.pcStart) >> 4
		bitmap[pc / 8] |= 1 << (pc % 8)
	}
	return data
}
```